crates/stages/src/stages/headers.rs

```diff
@@ -48,27 +34,33 @@ pub enum HeaderSyncMode {
 /// NOTE: This stage downloads headers in reverse. Upon returning the control flow to the pipeline,
 /// the stage checkpoint is not updated until this stage is done.
 #[derive(Debug)]
-pub struct HeaderStage<D: HeaderDownloader> {
+pub struct HeaderStage<Provider, Downloader: HeaderDownloader> {
+    /// Database handle.
+    provider: Provider,
```

Comment:
I like this a lot, no more fiddling with the db types directly,
we could also box this I think